src/logging/LazyLoggerFactory.ts

```diff
@@ -1,45 +1,110 @@
+  private readonly buffer: { logger: Logger; level: LogLevel; message: string; meta: any }[] = [];
+
+  public constructor(bufferSize = 1024) {
+    this.bufferSpaces = bufferSize;
   }
 
   public createLogger(label: string): Logger {
-    return new LazyLogger(this, label);
+    const wrapper = new WrappingLogger();
+    wrapper.logger = {
+      log: (level: LogLevel, message: string, meta: any): Logger =>
+        this.bufferLogEntry(wrapper, level, message, meta),
+    } as any;
+    this.wrappers.push({ wrapper, label });
+    return wrapper;
   }
 
-  public resetLoggerFactory(): void {
-    this.ploggerFactory = undefined;
+  private bufferLogEntry(logger: Logger, level: LogLevel, message: string, meta: any): Logger {
+    if (this.bufferSpaces > 0) {
+      // Buffer the message if spaces are still available
+      this.bufferSpaces -= 1;
+      if (this.bufferSpaces > 0) {
+        this.buffer.push({ logger, level, message, meta });
+      // Otherwise, generate a warning through a new logger
+      } else {
+        const wrapper = new WrappingLogger();
+        const bufferSize = this.buffer.length + 1;
+        this.wrappers.push({ wrapper, label: 'LazyLoggerFactory' });
+        this.buffer.push({
+          logger: wrapper,
+          level: 'warn',
+          message: `Memory-buffered logging limit of ${bufferSize} reached`,
+          meta: { bufferSize },
+        });
+      }
+    }
+    return logger;
+  }
+
+  /**
+   * Swaps all lazy loggers to new loggers from the given factory,
+   * and emits any buffered messages through those actual loggers.
+   */
+  public switch(loggerFactory: LoggerFactory): void {
+    // Instantiate an actual logger within every lazy logger
+    for (const { wrapper, label } of this.wrappers.splice(0, this.wrappers.length)) {
+      wrapper.logger = loggerFactory.createLogger(label);
+    }
+    // Emit all buffered log messages
+    for (const { logger, level, message, meta } of this.buffer.splice(0, this.buffer.length)) {
+      logger.log(level, message, meta);
+    }
+  }
+}
+
+/**
+ * Wraps around another {@link LoggerFactory} that can be set lazily.
+ * This is useful when objects are instantiated (and when they create loggers)
+ * before the logging system has been fully instantiated,
+ * as is the case when using a dependency injection framework such as Components.js.
+ *
+ * Loggers can be created even before a {@link LoggerFactory} is set;
+ * any log messages will be buffered and re-emitted.
+ */
+export class LazyLoggerFactory implements LoggerFactory {
+  private factory: LoggerFactory;
+
+  public constructor(options: { bufferSize?: number } = {}) {
+    this.factory = new TemporaryLoggerFactory(options.bufferSize);
   }
 
   public get loggerFactory(): LoggerFactory {
-    if (!this.ploggerFactory) {
-      throw new Error('No logger factory has been set. Can be caused by logger invocation during initialization.');
+    if (this.factory instanceof TemporaryLoggerFactory) {
```

Comment:
#464

Reply:
But https://github.com/CommunitySolidServer/CommunitySolidServer/issues/464 only holds for exported classes across files. This is a non-exported class within a file; there are no cases where this check fails.

Reply:
Fair enough. Was just thinking that it's just easiest for people if we have the rule "no `instanceof` at all"